### ISSUE TITLE
Bounds checks may sometimes be needed on D3D12

### ISSUE THREAD

**Issue by jimblandy** [gfx-rs/wgpu#4483]
Generally, HLSL requires bounds checks on indexing operations, so Naga's HLSL backend doesn't take a `BoundsCheckPolicies` option. However, in the [WGSL standards meeting today](https://docs.google.com/document/d/1U78rLoMBQQ5EkueX2FkNnDgEfY2G5ZlLV4wszqxojB4/edit#heading=h.1ofgl0uivvhc) a participant claimed that, under some circumstances, HLSL may not be doing those bounds checks:

> The no-writing OOB behaviour is limited to things with descriptors. (e.g. buffers, but not with things bound through root arguments in the root signature)  There’s no bounds checking on (ByteAddressBuffer, and StructuredBuffer as root arguments)  [Root sig 1.1](https://learn.microsoft.com/en-us/windows/win32/direct3d12/root-signature-version-1-1) has an option where a driver is allowed to move something from descriptor to root signature, and then lose the bounds checking.
See: [D3D12_DESCRIPTOR_RANGE_FLAGS (d3d12.h) - Win32 apps | Microsoft Learn](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/ne-d3d12-d3d12_descriptor_range_flags)
For: D3D12_DESCRIPTOR_RANGE_FLAG_DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS = 0x10000

**Comment by teoxoy** [comment]
Looks like we won't need extra guards in place after all https://github.com/gpuweb/gpuweb/issues/3893#issuecomment-1468753892

**Comment by jimblandy** [comment]
@teoxoy Was this fixed by #6431?

**Comment by teoxoy** [comment]
No, that one only covered the private, function and workgroup address spaces.

**Comment by teoxoy** [comment]
An update on this issue: We don't currently use Root Signature Version 1.1 (https://github.com/gfx-rs/wgpu/issues/6809) so, `D3D12_DESCRIPTOR_RANGE_FLAG_DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS` is not needed.

What we do need to do is inject bounds checks for dynamic buffers since those are root descriptors and not root descriptor tables.

**Comment by cwfitzgerald** [comment]
Idea!

Another possibility is to change how we desugar dynamically offset buffers. Instead of putting them in a root descriptor, we could have them and pass the offset as a root constant. This would work for storage buffers at least as we implement them as a ByteAddressBuffer. Constant buffers though we can't do this, but the bounds checking is much easier (all compile time sizes).

**Comment by teoxoy** [comment]
I think with that approach for dynamic storage buffers we could run into the max binding size limits?

**Comment by cwfitzgerald** [comment]
We cap the size of buffers to i32::MAX, and that's also the max_binding_size

**Comment by teoxoy** [comment]
Good point, I looked at how we can do it only for dynamic storage buffers and I'm not sure if it's possible because the BGL doesn't contain information about the kind of buffer uniform/storage.

**Comment by cwfitzgerald** [comment]
> I'm not sure if it's possible because the BGL doesn't contain information about the kind of buffer uniform/storage.

Huh? Yeah it does https://wgpu.rs/doc/wgpu/enum.BindingType.html#variant.Buffer.field.ty - we generate different code for them anyway right now.

**Comment by teoxoy** [comment]
Ah yeah, idk where I looked 😅 - I'll try that then.
